File: app/src/main/java/com/lessonscontrol/utils/FormatUtil.java
package com.lessonscontrol.utils;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.support.graphics.drawable.VectorDrawableCompat;
import android.text.TextUtils;

import java.security.InvalidParameterException;
import java.text.NumberFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import ca.antonious.materialdaypicker.MaterialDayPicker;

public abstract class FormatUtil {

    public static final String DATE_FORMAT = "dd/MM/yyyy";

    public static final String DEFAULT_VALUE = "";

    public static final String DAYS_FORMAT_SEPARATOR = "_";

    public static String convertDateToString(Long timeInMillis) {
        if (timeInMillis == null) {
            return DEFAULT_VALUE;
        }
        Date date = new Date(timeInMillis);
        SimpleDateFormat formatter = new SimpleDateFormat(DATE_FORMAT);
        return formatter.format(date);
    }

    public static String convertDoubleToMoney(Double value) {
        if (value == null) {
            return DEFAULT_VALUE;
        }
        NumberFormat formatter = NumberFormat.getCurrencyInstance();
        return formatter.format(value);
    }

    /**
     * Converts a list of {@link MaterialDayPicker} to a String, following the rule:
     * X_X..., X being the first three letters of that day of the week
     * Example: a list containing Sunday, Monday and Saturday would return SUN_MON_SAT
     *
     * @param days a list of {@link MaterialDayPicker}.
     * @return string in the specified format. null if the list is empty or null.
     */
    public static String convertMaterialDayPickerListToString(List<MaterialDayPicker.Weekday> days) {
        if (days == null || days.isEmpty()) {
            return null;
        }
        List<String> ids = new ArrayList<>();
        for (MaterialDayPicker.Weekday day : days) {
            ids.add(getWeekDayID(day));
        }
        return TextUtils.join(DAYS_FORMAT_SEPARATOR, ids);
    }

    public static List<MaterialDayPicker.Weekday> convertStringToMaterialDayPickerList(String selectedDays) {
        if (selectedDays == null || selectedDays.isEmpty()) {
            return null;
        }
        String[] listSelectedDaysIDs = selectedDays.split(DAYS_FORMAT_SEPARATOR);
        List<MaterialDayPicker.Weekday> days = new ArrayList<>();

        for (String id : listSelectedDaysIDs) {
            days.add(getWeekDayFromID(id));
        }

        return days;
    }

    public static String formatWeekDayIDsForDisplay(String ids, Resources resources, String contextPackageName) {
        if (ids == null || ids.isEmpty()) {
            return "";
        }
        List<String> formatted = new ArrayList<>();
        String[] individualIDsArray = ids.split(DAYS_FORMAT_SEPARATOR);
        for (String id : individualIDsArray) {
            formatted.add(resources.getString(resources.getIdentifier(id, "string", contextPackageName)));
        }
        return TextUtils.join(", ", formatted);
    }

    private static MaterialDayPicker.Weekday getWeekDayFromID(String id) {
        switch (id) {
            case "SUN":
                return MaterialDayPicker.Weekday.SUNDAY;
            case "MON":
                return MaterialDayPicker.Weekday.MONDAY;
            case "TUE":
                return MaterialDayPicker.Weekday.TUESDAY;
            case "WED":
                return MaterialDayPicker.Weekday.WEDNESDAY;
            case "THU":
                return MaterialDayPicker.Weekday.THURSDAY;
            case "FRI":
                return MaterialDayPicker.Weekday.FRIDAY;
            case "SAT":
                return MaterialDayPicker.Weekday.SATURDAY;
            default:
                throw new InvalidParameterException("Invalid weekday ID.");
        }
    }

    private static String getWeekDayID(MaterialDayPicker.Weekday day) {
        switch (day) {
            case SUNDAY:
                return "SUN";
            case MONDAY:
                return "MON";
            case TUESDAY:
                return "TUE";
            case WEDNESDAY:
                return "WED";
            case THURSDAY:
                return "THU";
            case FRIDAY:
                return "FRI";
            case SATURDAY:
                return "SAT";
            default:
                throw new InvalidParameterException("Invalid weekday.");
        }
    }

    public static Bitmap convertVectorDrawableToBitMap(VectorDrawableCompat vectorDrawable) {
        Bitmap bitmap = Bitmap.createBitmap(vectorDrawable.getIntrinsicWidth(),
                vectorDrawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        vectorDrawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        vectorDrawable.draw(canvas);
        return bitmap;
    }
}

File: app/src/main/java/com/lessonscontrol/utils/FormatUtil.kt
package com.lessonscontrol.utils

import android.content.res.Resources
import ca.antonious.materialdaypicker.MaterialDayPicker
import java.security.InvalidParameterException
import java.text.NumberFormat
import java.text.SimpleDateFormat
import java.util.*

object FormatUtil {

    private const val DATE_FORMAT = "dd/MM/yyyy"

    private const val DEFAULT_VALUE = ""

    private const val DAYS_FORMAT_SEPARATOR = "_"

    private const val DAYS_DISPLAY_SEPARATOR = ","

    fun convertDateToString(timeInMillis: Long?): String {
        if (timeInMillis == null) {
            return DEFAULT_VALUE
        }
        val date = Date(timeInMillis)
        val formatter = SimpleDateFormat(DATE_FORMAT)
        return formatter.format(date)
    }

    fun convertDoubleToMoney(value: Double?): String {
        if (value == null) {
            return DEFAULT_VALUE
        }
        val formatter = NumberFormat.getCurrencyInstance()
        return formatter.format(value)
    }

    /**
     * Converts a list of [MaterialDayPicker] to a String, following the rule:
     * X_X..., X being the first three letters of that day of the week
     * Example: a list containing Sunday, Monday and Saturday would return SUN_MON_SAT
     *
     * @param days a list of [MaterialDayPicker].
     * @return string in the specified format. null if the list is empty or null.
     */
    fun convertMaterialDayPickerListToString(days: List<MaterialDayPicker.Weekday>?): String {
        if (days == null || days.isEmpty()) {
            return DEFAULT_VALUE
        }
        return days.joinToString(separator = DAYS_FORMAT_SEPARATOR) { day -> getWeekDayID(day) }
    }

    fun convertStringToMaterialDayPickerList(selectedDays: String?): List<MaterialDayPicker.Weekday>? {
        if (selectedDays == null || selectedDays.isEmpty()) {
            return null
        }
        val listSelectedDaysIDs = selectedDays.split(DAYS_FORMAT_SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        return listSelectedDaysIDs.map { day -> getWeekDayFromID(day) }
    }

    fun formatWeekDayIDsForDisplay(ids: String?, resources: Resources, contextPackageName: String): String {
        if (ids == null || ids.isEmpty()) {
            return DEFAULT_VALUE
        }
        val individualIDsArray = ids.split(DAYS_FORMAT_SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        return individualIDsArray.joinToString(DAYS_DISPLAY_SEPARATOR) { id -> resources.getString(resources.getIdentifier(id, "string", contextPackageName)) }
    }

    private fun getWeekDayFromID(id: String): MaterialDayPicker.Weekday {
        return when (id) {
            "SUN" -> MaterialDayPicker.Weekday.SUNDAY
            "MON" -> MaterialDayPicker.Weekday.MONDAY
            "TUE" -> MaterialDayPicker.Weekday.TUESDAY
            "WED" -> MaterialDayPicker.Weekday.WEDNESDAY
            "THU" -> MaterialDayPicker.Weekday.THURSDAY
            "FRI" -> MaterialDayPicker.Weekday.FRIDAY
            "SAT" -> MaterialDayPicker.Weekday.SATURDAY
            else -> throw InvalidParameterException("Invalid weekday ID.")
        }
    }

    private fun getWeekDayID(day: MaterialDayPicker.Weekday): String {
        return when (day) {
            MaterialDayPicker.Weekday.SUNDAY -> "SUN"
            MaterialDayPicker.Weekday.MONDAY -> "MON"
            MaterialDayPicker.Weekday.TUESDAY -> "TUE"
            MaterialDayPicker.Weekday.WEDNESDAY -> "WED"
            MaterialDayPicker.Weekday.THURSDAY -> "THU"
            MaterialDayPicker.Weekday.FRIDAY -> "FRI"
            MaterialDayPicker.Weekday.SATURDAY -> "SAT"
            else -> throw InvalidParameterException("Invalid weekday.")
        }
    }
}
